### ISSUE TITLE
Label lat lon gridlines parallel to intersection angle at plot boundary

### ISSUE THREAD

**Issue by friedrichknuth** [SciTools/cartopy#1560]
### Description
Enhancement to orient lat lon gridline labels parallel to the gridlines as they intersect the plot boundary. Something like the labels seen here:

![plot](https://i.stack.imgur.com/Eichf.jpg)

[[source](https://www.the-cryosphere-discuss.net/tc-2020-80/tc-2020-80.pdf)]

### Code
```
import matplotlib.pyplot as plt
import cartopy.crs as ccrs

rotated_crs = ccrs.RotatedPole(pole_longitude=120.0, pole_latitude=70.0)

ax = plt.axes(projection=rotated_crs)
ax.set_extent((-6, 3, 48, 58), crs=ccrs.PlateCarree())
ax.coastlines(resolution='50m')

gl = ax.gridlines(draw_labels=True,x_inline=False, y_inline=False)
plt.show()
```

**Cartopy version**  
0.18.0

### Result
![here](https://i.stack.imgur.com/OIPiZ.png)

Notice the inconsistent orientation of gridline labels, which can get somewhat messy. It is possible to disable the automated orientation feature with `gl.rotate_labels = False`, however, I can't seem to find a method to align the label angle so that the text is parallel to the gridline as it intersects the plot boundary, as seen in the plot under the description.

### Additional Information
I think the logic for this might be here https://github.com/SciTools/cartopy/blob/47ec03f606d0833cc2aba4165efb8622fbe498bb/lib/cartopy/mpl/gridliner.py#L674 

This issue was briefly discussed here #https://github.com/SciTools/cartopy/pull/1089#issuecomment-488928836, but unsure if it has been addressed further.

**Comment by ocefpaf** [comment]
@dopplershift I believe that the behavior @friedrichknuth proposes here should be the default. It is quite odd to have different angle on the same side of the labels. What do you think?

**Comment by lacstorm** [comment]
This kind of gridline labels look much better!!

**Comment by icetianli** [comment]
By changing the `rotation` and `rotation_mode` parameters in `xlabel_style` and `ylabel_style` as follows can make the orientation of gridline labels in the sample region to be consistently horizontal:
```
gl.xlabel_style = {'size': 12, 'color': 'black', 'rotation': 0, 'rotation_mode': 'anchor'}
gl.ylabel_style = {'size': 12, 'color': 'black', 'rotation': 0, 'rotation_mode': 'anchor'}
```
This produces the map:
![image](https://user-images.githubusercontent.com/45267156/106366213-44e05800-6332-11eb-9007-13c82671cc10.png)

This at least makes the orientation of labels in the sample region to be consistent, but not sure how to make the text parallel to the gridline mentioned by @friedrichknuth. 

If change the map extent to a larger region, for example:
```
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
rotated_crs = ccrs.RotatedPole(pole_longitude=120.0, pole_latitude=70.0)
ax = plt.axes(projection=rotated_crs)
ax.set_extent((-10, 10, 50, 60), crs=ccrs.PlateCarree())
ax.coastlines(resolution='50m')
gl = ax.gridlines(draw_labels=True,x_inline=False, y_inline=False)
ticker_locator = 4
gl.xlocator = ticker.MultipleLocator(ticker_locator)
gl.ylocator = ticker.MultipleLocator(ticker_locator)
gl.xformatter = LongitudeFormatter()
gl.yformatter = LatitudeFormatter()
gl.xlabel_style = {'size': 12, 'color': 'black', 'rotation': 0, 'rotation_mode': 'anchor'}
gl.ylabel_style = {'size': 12, 'color': 'black', 'rotation': 0, 'rotation_mode': 'anchor'}
```
The above code produces map:
![image](https://user-images.githubusercontent.com/45267156/106366286-b91afb80-6332-11eb-8058-f61a94142b33.png)

It has an unwanted label `16E` on the right axis and its orientation is not horizontal anymore, looks like the `rotation` in `xlabel_style` and `ylabel_style` just can't change the orientation of all gridline labels at one time. 

Also, I would like to ask if there's a way to label only latitude or longitude on one side, instead of a mixture of both as the case of the right axis in the second figure.

**Comment by friedrichknuth** [comment]
To expand on @icetianli's question

_Is there's a way to label only latitude or longitude on one side, instead of a mixture of both as the case of the right axis in the second figure._

We can pass a list of fixed gridlines to draw and label, then turn off labels at the top, bottom, left, and right of the plot, e.g.

```
gl.xlocator = ticker.FixedLocator([-12, -8, -4, 0, 4, 8, 12,16])
gl.ylocator = ticker.FixedLocator([48, 52, 56, 60])

gl.top_labels = True
gl.bottom_labels = False
gl.left_labels = False
gl.right_labels = False
```

produces

![test](https://user-images.githubusercontent.com/10554254/113023711-c68a1180-913a-11eb-9581-befa2575ab9d.png)

however, we see that 8W and 12W still appear on the left side of the plot, despite `gl.left_labels = False`. 

I guess what is considered `left` vs `top` can get complicated with different projections, so perhaps there is a way to be explicit about which gridlines to draw vs label? 

Something like
```
gl.xlocator = mticker.FixedLocator([-12, -8, -4, 0, 4, 8, 12,16])
gl.xformatter = LongitudeFormatter([-4, 0, 4, 8, 12,16])
```
or
```
gl.xlocator = mticker.FixedLocator([-12, -8, -4, 0, 4, 8, 12,16])
lon_formatter = LongitudeFormatter()
lon_formatter.set_locs([-4, 0, 4, 8, 12,16])
labels = [lon_formatter(x) for x in [-4, 0, 4, 8, 12,16]]
gl.axes.set_xticklabels(labels)
```
... these examples don't change the result, but should help explain the idea. 

@ocefpaf is there a way to do this?

Ideally, one could be even more explicit and specify something like

```
gl.x_top_formatter = LongitudeFormatter([-4, 0, 4, 8, 12,16])
gl.x_bottom_formatter = LongitudeFormatter([-12, -8, -4, 0, 4, ])
```

I tried looking into creating a [secondary axis](https://stackoverflow.com/questions/19884335/matplotlib-top-bottom-ticks-different), but it doesn't seem to have all the necessary GeoAxes properties...

**Comment by QuLogic** [comment]
Closed by #1537
